### DISCUSSION TITLE
Struggling to make a search page for big database. IF the search has multiple optional forms, how do I query the db without hard coding every query permutation?

### DISCUSSION THREAD

**Discussion by bejp0a** [sqlpage/SQLPage#699] [Q&A]
Hello. I am fairly new to sql and very new to sqlpage. Apologies for any misunderstandings or omissions. 

I am hoping to use it as a simple search page (and ideally csv exporter) for a big database that I frequently need to query. Adding data or changing the data in the db is not in scope, I just need a search page that looks like this:

![Screenshot from 2024-11-17 12-28-40](https://github.com/user-attachments/assets/7c438425-8412-47b9-a541-c0bb0b951859)

That looks great, but it does not really do what I want under the hood. what I want is for sqlpage to run a query based on what feilds are filled out and what feilds are not. So when "Field1" and "Field5" are filled out it runs:

`SELECT * FROM "Table1" WHERE "field1" = "$Field1" AND "field5" = "$Field5";`

and if 1, 3 5 are filled out then query should be 

`SELECT * FROM "Table1" WHERE "field1" = "$Field1" AND "field3" = "$Field3" AND "field5" = "$Field5";`

but I cant figure out how to structure the sql in a way that avoids me having to write out all 64 permutations. I see in this [discussion](https://github.com/sqlpage/SQLPage/discussions/536) something that sort of works, but when I implement a slightly modified version  of the CASE test in the answer to that example the speed of the query takes a massive hit, to the point that its essentially unusable.

This sqlite database is more then 1TB and I don't have great hardware, but even so the difference the case test and the specific search is staggering. 

I see in other threads the author mentioning that arbitrary sql queries are not really supported, but I dont think thats what I want. I have a list of specific queries I want to run, I just dont want to have to type them all out (the number of feilds I want included will probably grow tbh)

here is my current index.sql, any advice or suggestions on how I can get this working would be greatly appreciated. 

```
SELECT 'form' as component, 'Search Tables' AS title, 'Search' as validate, 'GET' as method;

SELECT 
    'Table' as name,
    'Table to Search' as label,
    'select'  as type,
    TRUE      as required,
    1 as value,
    '[{"label": "Choose", "value": "1"}, {"label": "Table 1", "value": "2"}, {"label": "Table 2", "value": "3"}, {"label": "Table 3", "value": "4"}]' as options;

SELECT 'Field1' as label, 'Field1' as name, 'text' AS type, 5 as width;
SELECT 'Field2' as label, 'Field2' as name, 'text' AS type, 5 as width; 
SELECT 'Field3' as label, 'Field3' as name, 'text' AS type, 2 as width;
SELECT 'Field4' as label, 'Field4' as name, 'text' AS type, 5 as width;
SELECT 'Field5' as label, 'Field5' as name, 'text' AS type, 5 as width;
SELECT 'Field6' as label, 'Field6' as name, 'text' AS type, 2 as width; 

SELECT
        'table'         AS component,
        "Results"       AS title,
        TRUE            AS sort,
        TRUE            AS search;

SELECT * FROM "Table1" WHERE $Table = 2
AND
	CASE
		WHEN COALESCE($Field1,'') <> '' THEN 
			"Field1" = "$Field1" 
		ELSE "Field1" LIKE "*"
	END
AND
	CASE
		WHEN COALESCE($Field2,'') <> '' THEN 
			"Field2" = "$Field2" 
		ELSE "Field2" LIKE "*"
	END
LIMIT 1;

-- if it was not so slow, I would modify the above query to include all 6 feilds

SELECT * FROM "Table1" WHERE $Table = 3 AND "lname" = "$Field1" AND "field2" = "$Field2";

-- above works fantastic but I don't always use field1 and field 2. 

```

thanks in advance!

**Comment by lovasoa** [comment]
Hello and welcome to SQLPage!

I would love to know a little bit more about what you are building: a 1TB SQLite file sure isn't common!

About searching only the fields that were filled by the user, I think what you are looking for is a simple sequence of AND and OR operators: 

```sql
select * from table1
where
  ($field1 = '' OR field1 = $field1) AND
  ($field2 = '' OR field2 = $field2) AND
  ($field3 = '' OR field3 = $field3);
```

Or am I missing something?

**Reply by bejp0a** [reply]
Hello! thanks for the response. The problem with that or statement is the same is the CASE test, it does not use any of the indexs I've created and is thus far to slow to be usable. All of the feilds are indexed. When I write out this bash script to test the sql (sorry bout the spelling mistake I only noticed after writing this comment lol):

```
feild1='1'
feild2='2'
feild3='3'

sqlite3 bigdata.db <<EOS
EXPLAIN QUERY PLAN SELECT * FROM "Table1" WHERE feild1 = "$feild1" AND feild2 = "$feild2" AND feild3 = "$feild3";
EOS

sqlite3 bigdata.db <<EOS
EXPLAIN QUERY PLAN SELECT * from table1
where
  ($feild1 = '' OR feild1 = "$feild1") AND
  ($feild2 = '' OR feild2 = $feild2) AND
  ($feild3 = '' OR feild3 = $feild3);
EOS

```
When I run that in the terminal I get:

```
QUERY PLAN
`--SEARCH Table1 USING COVERING INDEX testindexf123 (feild1=? AND feild2=? AND feild3=?)
QUERY PLAN
`--SCAN Table1
```

So the first hard coded search uses the index but the AND OR sequence does not. The search needs to use an index for it to be useful, but I guess I dont know enough about sql / sqlite to know why this is happening, all the columns are indexed. the schema I used for this:

```
CREATE TABLE Table1(feild1 TEXT, feild2 TEXT, feild3 TEXT);
CREATE INDEX "testindexf1" ON "Table1" (feild1);
CREATE INDEX "testindexf2" ON "Table1" (feild2);
CREATE INDEX "testindexf3" ON "Table1" (feild3);
CREATE INDEX "testindexf123" ON "Table1" (feild1,feild2,feild3);
```
The database is a series of demographic csvs that I imported into a sqlite db that I am analysezing for a research project im presenting. I figured Id use sqlite because the fact that its based on a xyz.db file made the prospect of making short and long term backups much easier on my brain, also trying to get into android dev so figured I'd go with that.

**Reply by lovasoa** [reply]
The $variable syntax is specific to SQLPage, I don't think you can you it with the SQLite cli like that...

**Reply by lovasoa** [reply]
if you need more general query optimization advice, then I think you will get better answers on the SQLite forums than here. You can mention that you are using SQLite though sqlpage and even link to this issue over there. 

https://sqlite.org/forum/forum

**Reply by bejp0a** [reply]
> if you need more general query optimization advice, then I think you will get better answers on the SQLite forums than here. You can mention that you are using SQLite though sqlpage and even link to this issue over there.
> 
> https://sqlite.org/forum/forum

I asked over there and the two folks who responded suggested that it is better to have the queries built externally (presumibly by sqlpage) rather then trying to have one generic query. <https://sqlite.org/forum/forumpost/445e12149c445566>

I've started manually writing out all 64 possible permutations of the query, since you [so heavily discourage generating arbitrary queries, (I feel like generating "arbitrary" queries is not really what I want, but I'm not sure what else to call it )](https://github.com/sqlpage/SQLPage/discussions/518) (and to be honest even if I wanted to I'm not sure I could figure out how to, I know what the 64 queries should look like its starting to seem like it will be faster to just write them all out).

I've got about halfway through manually writing before I needed to switch to different task, Here is what I have so far, but its a huge monster that I really dont think is maintainable, especially since there are 64 potential filters but also several tables (currently 4 but more being added) that can be selected from drop down menu. 

But its what I am going to have to go with if I cant figure something else

```
-- $FILTER being the result of a test to see which fields filled out. 
-- if 1P, only one particular field is filled out
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 1 AND field1 = $field1;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 2 AND field2 = $field2;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 3 AND field3 = $field3;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 4 AND field4 = $field4;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 5 AND field5 = $field5;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 6 AND field6 = $field6;
--- if 2P, field1 and anothr
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 7 AND field1 = $field1 AND field2 = $field2;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 8 AND field1 = $field1 AND field3 = $field3;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 9 AND field1 = $field1 AND field4 = $field4;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 10 AND field1 = $field1 AND field5 = $field5;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 11 AND field1 = $field1 AND field6 = $field6;
--- if 3P, field1 and field2 and anothr
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 12 AND field1 = $field1 AND field2 = $field2 AND field3 = $field3;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 13 AND field1 = $field1 AND field2 = $field2 AND field4 = $field4;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 14 AND field1 = $field1 AND field2 = $field2 AND field5 = $field5;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 15 AND field1 = $field1 AND field2 = $field2 AND field6 = $field6;
--- if 3P, field1 and field3 and anothr
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 16 AND field1 = $field1 AND field3 = $field3 AND field4 = $field4;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 17 AND field1 = $field1 AND field3 = $field3 AND field5 = $field5;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 18 AND field1 = $field1 AND field3 = $field3 AND field6 = $field6;
--- if 3P, field1 and field4 and anothr
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 19 AND field1 = $field1 AND field4 = $field4 AND field5 = $field5;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 20 AND field1 = $field1 AND field4 = $field4 AND field6 = $field6;
--- if 3P, field1 and field5 and field6
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 21 AND field1 = $field1 AND field5 = $field5 AND field6 = $field6;
--- if 4P, field1 and field2 and field3 and anothr
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 22 AND field1 = $field1 AND field2 = $field2 AND field3 = $field3 AND field4 = $field4;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 23 AND field1 = $field1 AND field2 = $field2 AND field3 = $field3 AND field5 = $field5;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 24 AND field1 = $field1 AND field2 = $field2 AND field3 = $field3 AND field6 = $field6;
--- if 4P, field1 and field2 and field4 and anothr
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 25 AND field1 = $field1 AND field2 = $field2 AND field4 = $field4 AND field5 = $field5;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 26 AND field1 = $field1 AND field2 = $field2 AND field4 = $field4 AND field6 = $field6;
--- if 4P, field1 and field2 and field5 and field6
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 27 AND field1 = $field1 AND field2 = $field2 AND field5 = $field5 AND field6 = $field6;
--- if 5P, field1 and field2 and field3 and field4 and anothr
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 28 AND field1 = $field1 AND field2 = $field2 AND field3 = $field3 AND field4 = $field4 AND field5 = $field5;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 29 AND field1 = $field1 AND field2 = $field2 AND field3 = $field3 AND field4 = $field4 AND field6 = $field6;
--- if 5P, field1 and field2 and field3 and field5 and field6
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 30 AND field1 = $field1 AND field2 = $field2 AND field3 = $field3 AND field5 = $field5 AND field6 = $field6;
--- if 5P, field1 and field2 and field4 and field5 and field6
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 31 AND field1 = $field1 AND field2 = $field2 AND field4 = $field4 AND field5 = $field5 AND field6 = $field6;
--- ALL if 6P, field1 and field2 and and field3 and field4 and field5 and field6
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 32 AND field1 = $field1 AND field2 = $field2 AND field3 = $field3 AND field4 = $field4 AND field5 = $field5 AND field6 = $field6;

-- i think thats all field1, so rest shuld not have field1

--- if 2P, field2 and anothr
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 33 AND field2 = $field2 AND field3 = $field3;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 34 AND field2 = $field2 AND field4 = $field4;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 35 AND field2 = $field2 AND field5 = $field5;
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 36 AND field2 = $field2 AND field6 = $field6;
--- if 3P, field2 and field3 and anothr
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 37 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 38 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 39 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 40 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 41 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 42 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 43 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 44 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 45 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 46 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 47 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 48 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 49 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 50 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 51 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 52 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 53 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 54 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 55 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 56 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 57 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 58 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 59 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 60 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 61 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 62 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 63 AND
SELECT * FROM "table1" WHERE :Tablez = 2 AND $FILTER = 64 AND
``

**Reply by lovasoa** [reply]
Hi !
  I posted on the sqlite discussion thread. I ran a test, and I think the problem is a missing optimization in the sqlite3 query planner.
  The easiest solution is probably to switch to a database with a better query planner. I tested postgres on this example, and it does the required optimization and is able to use the required indices.

**Reply by lovasoa** [reply]
Alternatively, and probably more easily for you, you can add a first OR chain at the beginning of the query. It's not beautiful but it works, it makes the query planner use the indices:

```sql
SELECT * FROM test_table
WHERE
  (f1=$f1 OR f2=$f2 OR f3=$f3) AND
  ($f1 = '' OR f1=$f1) AND
  ($f2 = '' OR f2=$f2) AND
  ($f3 = '' OR f3=$f3);
```

<details>

```
sqlite> EXPLAIN QUERY PLAN 
SELECT * FROM test_table 
WHERE 
   (field1='adabfeffde' OR field2 = '' OR field3 = 'aaaaaaaaaa') AND 
   (field1='adabfeffde' OR ''='adabfeffde') AND 
   (field2='' OR ''='') AND 
   (field3='aaaaaaaaaa' OR ''='aaaaaaaaaa')
;
QUERY PLAN
`--MULTI-INDEX OR
   |--INDEX 1
   |  `--SEARCH test_table USING INDEX idx_field1 (field1=?)
   |--INDEX 2
   |  `--SEARCH test_table USING INDEX idx_field2 (field2=?)
   `--INDEX 3
      `--SEARCH test_table USING INDEX idx_field3 (field3=?)
```

</details>

**Reply by heejit** [reply]
if it is a text field you can concat all the field and search like this ->
`concat(fd1, fd2) like %input%' or concat_ws('', fd1, fd2)`
you can use one input only instead of multiple

**Reply by lovasoa** [reply]
@heejit : I don't think this will use the field indices (which is @bejp0a problem).

**Reply by heejit** [reply]
yes it will not use indices

**Reply by heejit** [reply]
or create a new field with concat all the fields and create index on it

**Reply by heejit** [reply]
or use generated column and create index on it.

**Reply by bejp0a** [reply]
I ended up using a spreadsheet and ctl h to make a list of every possible query based on the availible forms, as well as a case test to print out a number from 1-63. 

But now I cant figure out how to get sqlpage (or indeed sqlite) to perform a specific query based on a case test. IE, 

SELECT CASE
  WHERE $FILTER = 1 THEN (SELECT * FROM table WHERE f1 = $f1)
  WHERE $FILTER = 2 THEN (SELECT * FROM table WHERE f1 = $f1 AND f2 = $f2)
ELSE (SELECT '0')

returns an error with "error returned from database: (code: 1) sub-select returns 33 columns - expected 1", and I cant figure out how to work around it. I guess CASE is only supposed to return one value? Is there a better way to get sqlpage to perform a specific query that I have hard coded based on the value of a variable? not an arbitrary query, it just needs to pick from a list of predifines set of queries that I know are safe. 

I attempted to get your proposed solution to work as well, but again its simply to slow to be usable. takes less then a second to run SELECT * FROM table WHERE f1=$f1 AND f2=$f2 AND f3=$f3 but yours is taking more then a minute before I hit ctl c. It does say its using index's but in a weird order, perhapes becuase I built the index wrong, so I'm recreating the db and will attempt to run your example query (dont have space to add any more indexes or to vacume the db, since vaccuming creats a full copy and no space for that) and report back.

EDIT: ah wait, maybe I needed to try [something like this ](https://dba.stackexchange.com/questions/158141/how-to-return-more-then-1-value-with-case-statement)

```
SELECT * FROM table 
WHERE (WHERE $FILTER = 1 AND f1 = $f1)
     OR (WHERE $FILTER = 1 AND f1 = $f1  AND f2 = $f2)
     OR (WHERE $FILTER = 1 AND f1 = $f1  AND f2 = $f2 AND f3 = $f3)
      ) ;
```

will give that a go when the db is recreated

**Reply by lovasoa** [reply]
Hi ! As I stated above, the best solution is probably not to create many different database queries.

This solves your problem, is concise, and uses indices:

```sql
SELECT * FROM test_table
WHERE
  (f1=$f1 OR f2=$f2 OR f3=$f3) AND
  ($f1 = '' OR f1=$f1) AND
  ($f2 = '' OR f2=$f2) AND
  ($f3 = '' OR f3=$f3);
```

**Reply by lovasoa** [reply]
Ok, I just read your updated answer above. If you really need to run dynamic sql queries, you can try the following: https://github.com/sqlpage/SQLPage/discussions/518#discussioncomment-10210484
